### PR TITLE
fflas-ffpack: Add darwin support

### PR DESCRIPTION
**Pull request by jbaum98**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jbaum98** [comment]
This actually isn't working: it stills fails tests, but they weren't being run before because `doCheck` was set to false even though `doInstallCheck` was set to true. Will reopen when I fix it.

**Comment by jbaum98** [comment]
The problem seems to be with `openblas` on darwin. It works with `blas`. I will also try with `atlas`.

Is there a known issue with `openblas` on darwin?

**Comment by timokau** [comment]
Not that I know of, what is the error?

Common practice would be to leave the PR open but add a [WIP] prefix to the title.

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (timokau) [inline pkgs/development/libraries/fflas-ffpack/default.nix]: Same as the other tickets, please expand and report upstream. Does the same issue actually affect all 3 packages? I could understand linbox and givaro since they are from the same people, but this seems weird.

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (timokau) [inline pkgs/development/libraries/fflas-ffpack/default.nix]: Not sure what github is telling me here. Are you adding whitespace at the start of the line / changing spaces to tabs or something?

**Comment by timokau** [comment]
There is [this](https://salsa.debian.org/science-team/sagemath/commit/c89e17cb9cc05d9caa0ae5342ba7511aa8d59b6c), but they don't say anything about darwin. Might be relevant, might not be.

**Comment by timokau** [comment]
Still I'd avoid switching to atlas if possible.

**Comment by jbaum98** [comment]
So I should just use regular blas instead of openblas over atlas?

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (jbaum98) [reply pkgs/development/libraries/fflas-ffpack/default.nix]: No this was my mistake, I saw a segfault and assumed it was the same problem, but it was actually openblas. But then it wasn't running the tests so it was looking like it was working.

**Comment by timokau** [comment]
I don't think there is such a thing as regular blas (might be wrong). blas just defines an interface and openblas and atlas are the two main implementations. It should stay with openblas if possible. openblas seems generally preferred, has more features and is better supported on nix (more optimizations).

What error are you getting?

**Comment by jbaum98** [comment]
There is a package called blas that I used that works. It's segfaulting on
a test, and according to my debugger it's happening in a call to an
openblas method. Will post more details soon.

**Comment by timokau** [comment]
Oh, didn't know about that one. Probably that is the one that started the standard. I always thought the first one was proprietary for some reason.

Apparently [this](https://github.com/xianyi/OpenBLAS/issues/1641) is the upstream issue that caused debian to temporarily switch to atlas. You could try if downgrading openblas to 0.3.1 fixes the issue.

**Comment by timokau** [comment]
From your history it looks like you're trying to get sage to build on darwin, is that right?

**Comment by jbaum98** [comment]
Yup, you figured me out.

**Comment by timokau** [comment]
Awesome! I'm the sage maintainer. If you need help or review, ping me.

**Comment by jbaum98** [comment]
Okay here's the situation. There are three (maybe four) `BLAS` libraries we could use: `openblas`, `blas`, `Accelerate`, and maybe `atlas`. It seems like `openblas` is what we'd like to use, but it doesn't work on darwin:
```
==============================================
   FFLAS-FFPACK 2.3.2: tests/test-suite.log
==============================================

# TOTAL: 33
# PASS:  11
# SKIP:  0
# XFAIL: 0
# FAIL:  22
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test-fger
===============

FAIL test-fger (exit status: 139)

FAIL: test-ftrsv
================

Checking with Modular<double> mod 36289873
FAIL test-ftrsv (exit status: 139)

FAIL: test-ftrtri
=================

Checking with Modular<double> mod 131
Checking FTRTRI_Lower_Unit....................PASSED (0.031471)
Checking FTRTRI_Upper_Unit....................PASSED (0.017059)
Checking FTRTRI_Lower_NonUnit.................PASSED (0.023665)
Checking FTRTRI_Upper_NonUnit.................PASSED (0.021203)
Checking with Modular<double> mod 13
Checking FTRTRI_Lower_Unit....................PASSED (0.018172)
Checking FTRTRI_Upper_Unit....................PASSED (0.020317)
Checking FTRTRI_Lower_NonUnit.................PASSED (0.016548)
Checking FTRTRI_Upper_NonUnit.................PASSED (0.00784)
Checking with Modular<double> mod 33811
Checking FTRTRI_Lower_Unit....................PASSED (0.004757)
Checking FTRTRI_Upper_Unit....................PASSED (0.004347)
Checking FTRTRI_Lower_NonUnit.................PASSED (0.004994)
Checking FTRTRI_Upper_NonUnit.................PASSED (0.007178)
Checking with ModularBalanced<double> mod 131
FAIL test-ftrtri (exit status: 139)

FAIL: test-ftrmv
================

Checking with Modular<double> mod 646771
Checking FTRMV_Lower_NoTrans_Unit............PASSED (0.004181)
FAIL test-ftrmv (exit status: 139)

FAIL: test-ftrsm
================

Checking with Modular<double> mod 1547437
FAIL test-ftrsm (exit status: 139)

FAIL: test-ftrsm-check
======================

FAIL test-ftrsm-check (exit status: 139)

FAIL: test-ftrmm
================

Checking with Modular<double> mod 61
FAIL test-ftrmm (exit status: 139)

FAIL: test-fgemm
================

Checking ..........................Modular<double> mod 1747 ...  ** On entry to DGEMM  parameter number 10 had an illegal value
 ** On entry to DGEMM  parameter number 10 had an illegal value
 ** On entry to DGEMM  parameter number 10 had an illegal value
 ** On entry to DGEMM  parameter number 10 had an illegal value
 ** On entry to DGEMM  parameter number 10 had an illegal value
 ** On entry to DGEMM  parameter number 10 had an illegal value
FAIL
a   :1, b   : 0
m   :16, n   : 44, k   : 36
ldA :48, ldB : 47, ldC : 45
Error C[0,0]=813 D[0,0]=202
Error C[0,1]=205 D[0,1]=612
Error C[0,2]=256 D[0,2]=1713
Error C[0,3]=102 D[0,3]=494
Error C[0,4]=872 D[0,4]=631
Error C[0,5]=165 D[0,5]=699
Error C[0,6]=1299 D[0,6]=1459
Error C[0,7]=246 D[0,7]=70
Error C[0,8]=1737 D[0,8]=1069
Error C[0,9]=592 D[0,9]=1088
Error C[0,10]=401 D[0,10]=637
Error C[0,11]=396 D[0,11]=1058
Error C[0,12]=1587 D[0,12]=848
Error C[0,13]=1498 D[0,13]=838
Error C[0,14]=1695 D[0,14]=1059
Error C[0,15]=196 D[0,15]=974
Error C[0,16]=290 D[0,16]=1518
Error C[0,17]=1634 D[0,17]=1479
Error C[0,18]=213 D[0,18]=1132
Error C[0,19]=769 D[0,19]=1695
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
FAILED
FAIL test-fgemm (exit status: 1)

FAIL: test-fgemm-check
======================

FAIL test-fgemm-check (exit status: 139)

FAIL: test-lu
=============

FAIL test-lu (exit status: 139)

FAIL: test-pluq-check
=====================

FAIL test-pluq-check (exit status: 139)

FAIL: test-fsyrk
================

Checking with Modular<double> mod 5
FAIL test-fsyrk (exit status: 139)

FAIL: test-fsytrf
=================

FAIL test-fsytrf (exit status: 139)

FAIL: test-invert-check
=======================

 -q 131071 -n 1000 -i 3 -s 1534291159349551
FAIL test-invert-check (exit status: 139)

FAIL: test-rankprofiles
=======================

FAIL test-rankprofiles (exit status: 139)

FAIL: test-det-check
====================

init: 0.00863194s (0.02807 cpu) [1]
FAIL test-det-check (exit status: 139)

FAIL: test-echelon
==================

FAIL test-echelon (exit status: 139)

FAIL: test-charpoly
===================

FAIL test-charpoly (exit status: 139)

FAIL: test-charpoly-check
=========================

FAIL test-charpoly-check (exit status: 139)

FAIL: test-minpoly
==================

FAIL test-minpoly (exit status: 139)

FAIL: test-solve
================

FAIL test-solve (exit status: 139)

FAIL: test-fgemv
================

FAIL test-fgemv (exit status: 139)

============================================================================
Testsuite summary for FFLAS-FFPACK 2.3.2
============================================================================
# TOTAL: 33
# PASS:  11
# SKIP:  0
# XFAIL: 0
# FAIL:  22
# XPASS: 0
# ERROR: 0
```
I looked at `test-fger` in `lldb`, and it seemed like the problem was within a call to an `openblas` function `cblas_dger`:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x0000000102ec6e50 libopenblasp-r0.3.1.dylib`daxpy_kernel_8 + 16
libopenblasp-r0.3.1.dylib`daxpy_kernel_8:
->  0x102ec6e50 <+16>: vmovups (%rdx,%rax,8), %ymm12
    0x102ec6e55 <+21>: vmovups 0x20(%rdx,%rax,8), %ymm13
    0x102ec6e5b <+27>: vmovups 0x40(%rdx,%rax,8), %ymm14
    0x102ec6e61 <+33>: vmovups 0x60(%rdx,%rax,8), %ymm15
Target 0: (test-fger) stopped.
(lldb) up
frame #1: 0x0000000102ec69cb libopenblasp-r0.3.1.dylib`daxpy_k_HASWELL + 91
libopenblasp-r0.3.1.dylib`daxpy_k_HASWELL:
    0x102ec69cb <+91>:  cmpq   %r14, %rbx
    0x102ec69ce <+94>:  jge    0x102ec6e27               ; <+1207>
    0x102ec69d4 <+100>: movsd  -0x38(%rbp), %xmm0        ; xmm0 = mem[0],zero
    0x102ec69d9 <+105>: movq   %r14, %rax
(lldb) up
frame #2: 0x0000000102ece2e9 libopenblasp-r0.3.1.dylib`dger_k_HASWELL + 185
libopenblasp-r0.3.1.dylib`dger_k_HASWELL:
    0x102ece2e9 <+185>: movsd  -0x30(%rbp), %xmm1        ; xmm1 = mem[0],zero
    0x102ece2ee <+190>: movq   %r12, %rcx
    0x102ece2f1 <+193>: movq   %r15, %r12
    0x102ece2f4 <+196>: addq   $0x20, %rsp
(lldb) up
frame #3: 0x00000001001fcadb libopenblasp-r0.3.1.dylib`cblas_dger + 507
libopenblasp-r0.3.1.dylib`cblas_dger:
    0x1001fcadb <+507>: jmp    0x1001fcafd               ; <+541>
    0x1001fcadd <+509>: movq   %r13, %rdi
    0x1001fcae0 <+512>: movq   %r12, %rsi
    0x1001fcae3 <+515>: movq   %rcx, %rdx
(lldb) up
test-fger was compiled with optimization - stepping may behave oddly; variables may not be available.
frame #4: 0x000000010000ce26 test-fger`FFLAS::fger(F=<unavailable>, M=44, N=36, alpha=<unavailable>, x=0x0000000104701910, incx=1, y=0x0000000104701af0, incy=<unavailable>, A=<unavailable>, lda=<unavailable>, H=0x00007ffeefbfe538) at fflas_fger.inl:185 [opt]
   182 	#ifdef __FFLASFFPACK_OPENBLAS_NUM_THREADS
   183 			openblas_set_num_threads(__FFLASFFPACK_OPENBLAS_NUM_THREADS);
   184 	#endif
-> 185 			cblas_dger( CblasRowMajor, (int)M, (int)N, alpha, x, (int)incx, y, (int)incy, A, (int)lda );
   186 		}
   187 	
   188 		template<class Field>
```

`Accelerate` is made just for mac so it should be well-tuned, so that seems like it's our second choice. Unfortunately it has one failing test:
```
==============================================
   FFLAS-FFPACK 2.3.2: tests/test-suite.log
==============================================

# TOTAL: 33
# PASS:  32
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test-ftrsm
================

Checking with Modular<double> mod 10103
Checking FTRSM_Lower_Left_NoTrans_Unit............PASSED (0.003281)
Checking FTRSM_Upper_Left_NoTrans_Unit............PASSED (0.002883)
libc++abi.dylib: terminating with uncaught exception of type FailureFgemmCheck
Checking FTRSM_Lower_Left_Trans_Unit..............FAIL test-ftrsm (exit status: 134)
```
This may or may not be a bug, because the [README](https://github.com/linbox-team/fflas-ffpack/blob/master/README.md) says it is, but [the issue it links to](https://github.com/linbox-team/fflas-ffpack/issues/29) is closed, so I'm confused.

So, then, we settle for `blas` and it build no problem. I haven't had a chance to get `atlas` building to see if that'll work.

**Comment by timokau** [comment]
Did you try to build with an older version of `openblas`?

**Comment by jbaum98** [comment]
Building with the newer version of `openblas`, `0.3.2`, was the same. I'll try with `0.3.0` now. Is there a specific version we think made a breaking change?

**Comment by jbaum98** [comment]
I tried with a newer `0.3.2` and older `0.3.0`, all the same. Should I go older?

**Comment by timokau** [comment]
Oh I thought we already updated to `0.3.2` in nixpkgs and that might be responsible. Then I'm out of ideas, we should ask upstream. If you don't want to wait for an upstream reply, I'd be fine with conditionally (on darwin) using whatever blas implementation works for now.

**Comment by jbaum98** [comment]
Okay so regular `blas`. I will say that's what same seems to use. How should I ask upstream?

**Comment by timokau** [comment]
> same

You mean sage? They use openblas. I would guess (going from that I haven't heard of plain `blas` before) that atlas is probably more performant. But either way, working on darwin is better than not working on darwin.

Just add an issue to fflas-ffpack's github page explaining that the build fails on darwin with openblas while it succeeds on linux with openblas and on darwin with another blas. Link to this issue and especially your error logs. They may have insight that is obvious to them but not to us. Then you can link the upstream issue to the place where you conditionally use another blas for darwin.

**Comment by timokau** [comment]
No, they actually provide `.pc` files for `pkg-config` to make `openblas` count as a `cblas` library. Its confusing. So `pkg-config --libs cblas` should actually return the path to openblas (just as `blas`, `lapack` or `openblas` would).

For details see `build/pkgs/openblas/write_pc_file.py` in the sage repo. In nix there is `openblas-pc.nix` for the same purpose (although I'm not actually sure right now if that is still necessary, I'll have to re-check that some time).

**Comment by jbaum98** [comment]
Can I use that to handle the `--with-blas-libs` in a uniform way? Right now I have to explicitly pass it in `all-packages.nix`.

**Comment by jbaum98** [comment]
Maybe `sage` just doesn't run tests, and that's why they can get away with using `openblas` on darwin and we can't?

**Comment by timokau** [comment]
> Can I use that to handle the --with-blas-libs in a uniform way? Right now I have to explicitly pass it in all-packages.nix.

Good question, I don't think you can. You could just make the value dependent on `isDarwin`.

> Maybe sage just doesn't run tests, and that's why they can get away with using openblas on darwin and we can't?

Again, its a bit confusing :grinning: 
They do run the tests on some dependencies sometimes. Specifically, they run the dependencies on all the packages where a `spkg-check` (instructions on how to run the tests are to be run), but not by default. Since there is no `spkg-check` in `build/pkgs/fflas_ffpack`, it looks like they don't run the tests for this package.

**Comment by timokau** [comment]
So if you want to double check what the reason for the tests is, you could write an `spkg-check` for sage's fflas-ffpack and try to build sage with tests from there. But that will take a long time and is quite some effort that will likely have the same failure as here. I'd try the upstream report first. Often times an issue packagers spend ages to debug is obvious to people familiar with the internals.

**Comment by timokau** [comment]
If you'd like to keep it the way it is now, please add a comment to the package file somewhere explaining that openblas with osx causes issues, linking to the upstream issue and mentioning that `blas` is set conditionally in `all-packages.nix`. If we only have the blas conditional in `all-packages.nix` it is likely to be forgotten, even if the issue is fixed upstream in the future.

After that if you're happy with it and tests pass on both platforms, I'd say its good to go.

**Comment by timokau** [comment]
For the record, I did ask [on irc](https://logs.nix.samueldr.com/nixos-dev/2018-08-16#1482703;) if there is a better way for the lib name, but apparently there is not.

File: pkgs/top-level/all-packages.nix
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: I think this is a bit confusing since `blas` is supposed to be a package. Adding a linkName as a passthru attribute to the packages and using `blas.linkName or "blas"` instead seems much nicer.

Comment (jbaum98) [reply pkgs/top-level/all-packages.nix]: Oh, a passthru attribute is exactly what we want. Didn't know about that!

Comment (timokau) [reply pkgs/top-level/all-packages.nix]: As always, please add a comment about the purpose of the `linkName` attribute. Otherwise these things tend to stay around forever even after nobody uses them anymore because nobody knows what it was good for.

File: pkgs/top-level/all-packages.nix
Comment (timokau) [inline pkgs/top-level/all-packages.nix]: You can just add it directly to the openblas and blas packages. `passthru` won't cause a rebuild.

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (timokau) [inline pkgs/development/libraries/fflas-ffpack/default.nix]: Not 100% sure on this but you should be able to just do `blas.linkName`.

Comment (LnL7) [reply pkgs/development/libraries/fflas-ffpack/default.nix]: Indeed

File: pkgs/top-level/all-packages.nix
Comment (jbaum98) [reply pkgs/top-level/all-packages.nix]: Ohh you're saying I should modify them to add the `passthru` attribute in `all-packages.nix`.

File: pkgs/top-level/all-packages.nix
Comment (LnL7) [inline pkgs/top-level/all-packages.nix]: I think I didn't explain theis properly, there's no need for overrides in all-packages.nix. Add the overrides eg. here, similar to meta this won't cause a rebuild when changed. https://github.com/NixOS/nixpkgs/blob/55c4572dd1fae7043b1697116007d0f35edbcb64/pkgs/development/libraries/science/math/blas/default.nix#L55

Comment (timokau) [reply pkgs/top-level/all-packages.nix]: So you can directly add this to `pkgs/development/libraries/sciene/math/{open,}blas/default.nix`.

Comment (jbaum98) [reply pkgs/top-level/all-packages.nix]: Yeah sorry, makes sense now.

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (timokau) [inline pkgs/development/libraries/fflas-ffpack/default.nix]: Why did you remove those?

**Comment by timokau** [comment]
As soon as you think this is ready, please remove the wip tag.

**Review by LnL7 — APPROVED** [review]
Looks good now.

File: pkgs/development/libraries/fflas-ffpack/default.nix
Comment (jbaum98) [reply pkgs/development/libraries/fflas-ffpack/default.nix]: What do you mean?

Comment (timokau) [reply pkgs/development/libraries/fflas-ffpack/default.nix]: Never mind, I must've mis-read the github diff. I thought you removed a bunch of `--disable-X`.

**Comment by timokau** [comment]
@GrahamcOfBorg build fflas-ffpack

**Comment by timokau** [comment]
I've reported this issue at sage upstream, maybe somebody there can help: https://trac.sagemath.org/ticket/26130

**Pull request merged**
